src/esp/gfx/DebugLineRender.cpp

```diff
@@ -0,0 +1,314 @@
+  // remap alpha (opacity). This is an approximation.
+  constexpr float exponent = 2.f;
+  return Mn::Color4(src.r(), src.g(), src.b(), std::pow(src.a(), exponent));
+}
+
+// return false if segment is entirely clipped
+bool scissorSegmentToOutsideCircle(Mn::Vector3* pt0,
+                                   Mn::Vector3* pt1,
+                                   const Mn::Vector3& center,
+                                   float radius) {
+  float dist0 = (*pt0 - center).length();
+  float dist1 = (*pt1 - center).length();
+  if (dist0 < radius) {
+    if (dist1 < radius) {
+      return false;
+    }
+    const float lerpFraction = (radius - dist0) / (dist1 - dist0);
+    CORRADE_INTERNAL_ASSERT(lerpFraction >= 0.f && lerpFraction <= 1.f);
+    Mn::Vector3 clippedPt = Mn::Math::lerp(*pt0, *pt1, lerpFraction);
+    *pt0 = clippedPt;
+    return true;
+  }
+
+  if (dist1 < radius) {
+    const float lerpFraction = (radius - dist1) / (dist0 - dist1);
+    CORRADE_INTERNAL_ASSERT(lerpFraction >= 0.f && lerpFraction <= 1.f);
+    Mn::Vector3 clippedPt = Mn::Math::lerp(*pt1, *pt0, lerpFraction);
+    *pt1 = clippedPt;
+    return true;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+DebugLineRender::DebugLineRender()
+    : _glResources{std::make_unique<GLResourceSet>()} {
+  _glResources->mesh.addVertexBuffer(_glResources->buffer, 0,
+                                     Mn::Shaders::VertexColor3D::Position{},
+                                     Mn::Shaders::VertexColor3D::Color4{});
+}
+
+void DebugLineRender::releaseGLResources() {
+  _glResources = nullptr;
+}
+
+void DebugLineRender::drawLine(const Mn::Vector3& from,
+                               const Mn::Vector3& to,
+                               const Mn::Color4& fromColor,
+                               const Mn::Color4& toColor) {
+  VertexRecord v1{from, remapAlpha(fromColor)};
+  VertexRecord v2{to, remapAlpha(toColor)};
+  arrayAppend(_verts, {v1, v2});
+}
+
+void DebugLineRender::setLineWidth(float lineWidth) {
+  // This is derived from experiments with glLineWidth on Nvidia hardware.
+  const float maxLineWidth = 20.f;
+  if (lineWidth > maxLineWidth) {
+    LOG(WARNING) << "DebugLineRender::setLineWidth: requested lineWidth of "
+                 << lineWidth
+                 << " is greater than "
+                    "max supported width of "
+                 << maxLineWidth;
+    lineWidth = maxLineWidth;
+  }
+  _internalLineWidth = lineWidth / 2;  // see also DebugLineRender::flushLines
+}
+
+void DebugLineRender::flushLines(const Magnum::Matrix4& camMatrix,
+                                 const Magnum::Matrix4& projMatrix,
+                                 const Magnum::Vector2i& viewport) {
+  CORRADE_ASSERT(_glResources,
+                 "DebugLineRender::flushLines: no GL resources; see "
+                 "also releaseGLResources", );
+
+  bool doToggleBlend = !glIsEnabled(GL_BLEND);
```

Comment:
@mosra @bigbike I don't love calling `glIsEnabled` but I didn't see a Magnum getter.

What I really need here is a way to "push" a render state and later pop it (then I wouldn't need the getter).

I'm open to other suggestions.

Reply:
Yeah sorry, a nasty pain point. 

This should _ideally_ eventually be handled by a scoped renderer state, where you enable what you need and Magnum's state tracker then takes care of cleaning that up again at the end of scope, _ideally_ also not calling into GL more than necessary. Practically we'll probably switch to Vulkan sooner than I get to implement such a feature.